hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/NamenodeHeartbeatService.java

```diff
@@ -399,13 +356,58 @@ private static String getNnHeartBeatServiceName(String nsId, String nnId) {
         (nnId == null ? "" : " " + nnId);
   }
 
+  /**
+   * Get the namespace information for a Namenode via RPC and add them to the report.
+   * @param serviceURI Server address of the Namenode to monitor.
+   * @param report Namenode status report updating with namespace information data.
+   */
+  private void updateNameSpaceInfoParameters(
+      URI serviceURI, NamenodeStatusReport report) throws IOException {
+    try {
+      if (this.namenodeProtocol == null) {
+        this.namenodeProtocol = NameNodeProxies.createProxy(this.conf, serviceURI,
+            NamenodeProtocol.class).getProxy();
+      }
+      if (namenodeProtocol != null) {
+        NamespaceInfo info = namenodeProtocol.versionRequest();
+        if (info != null) {
+          report.setNamespaceInfo(info);
+        }
+      }
+    } catch (IOException e) {
+      this.namenodeProtocol = null;
+      throw e;
+    }
+  }
+
+  /**
+   * Get the safemode information for a Namenode via RPC and add them to the report.
+   * @param serviceURI Server address of the Namenode to monitor.
+   * @param report Namenode status report updating with safemode information data.
+   */
+  private void updateSafeModeParameters(URI serviceURI, NamenodeStatusReport report) {
+    try {
+      if (this.clientProtocol == null) {
+        this.clientProtocol = NameNodeProxies
+            .createProxy(this.conf, serviceURI, ClientProtocol.class)
+            .getProxy();
+      }
+      if (clientProtocol != null) {
+        boolean isSafeMode = clientProtocol.setSafeMode(SafeModeAction.SAFEMODE_GET, false);
+        report.setSafeMode(isSafeMode);
+      }
+    } catch (Exception e) {
+      LOG.error("Cannot fetch safemode state for {}", getNamenodeDesc(), e);
```

Comment:
Exception why not throw?

Reply:
The original logic just ignores this exception, so I just warp them into a method, maybe safe mode status of namenode is not important for RBF.

If you have some corner cases about it, please share with me, and I will fix it.

Reply:
The reason I have this question is because in the updateNameSpaceInfoParameters function, you throw an exception

```
private void updateNameSpaceInfoParameters(
      URI serviceURI, NamenodeStatusReport report) throws IOException {
    try {
      if (this.namenodeProtocol == null) {
        this.namenodeProtocol = NameNodeProxies.createProxy(this.conf, serviceURI,
            NamenodeProtocol.class).getProxy();
      }
    ...
    } catch (IOException e) {
      this.namenodeProtocol = null;
      throw e;
    }
  }
```

Reply:
updateNameSpaceInfoParameters and updateSafeModeParameters  seem to have similar functions, but the exception handling is inconsistent, which makes me have some doubts, just personal opinion.

Reply:
Yes, the exception handling is inconsistent. 

NameSpaceInfo is very important for RBF to identify one NameNode, so if there are some exceptions, we don't need to get any other information from namenode, so throw exception, just print some warn log and let the report invalid. 

Safe mode status of NameNode is useless for RBF to identify one NameNode, so if there are some exceptions, we can ignore it and can try to get other informations from Namenode.

Of course,  @goiri can give the most authoritative explanation.

Reply:
I think is fair to have different behaviors.
However, we should add that difference to the javadoc and justify it.

Reply:
@ZanderXu I see your change, thank you very much, but I would like to ask if we can unify these 2 ways, like throwing exceptions for both?

Reply:
Oh, I saw this comment. @slfan1989 About this, I think is fair to have different behaviors, because there are indeed different behaviors.